picoquic/sender.c

```diff
@@ -2043,25 +2050,29 @@ int picoquic_prepare_packet_server_init(picoquic_cnx_t* cnx, picoquic_path_t * p
             packet->length = length;
 
         }
-        else  if ((length = picoquic_retransmit_needed(cnx, pc, path_x, current_time, next_wake_time, packet, send_buffer_max, &is_cleartext_mode, &header_length)) > 0) {
+        else  if ((cnx->initial_validated || cnx->initial_repeat_needed) && 
+        (length = picoquic_retransmit_needed(cnx, pc, path_x, current_time, next_wake_time, packet, send_buffer_max, &is_cleartext_mode, &header_length)) > 0) {
             /* Set the new checksum length */
             checksum_overhead = picoquic_get_checksum_length(cnx, is_cleartext_mode);
-            /* Check whether it makes sens to add an ACK at the end of the retransmission */
-            ret = picoquic_prepare_ack_frame(cnx, current_time, pc, &bytes[length],
-                send_buffer_max - checksum_overhead - length, &data_bytes);
-            if (ret == 0) {
-                length += data_bytes;
-                packet->length = length;
-            }
-            else if (ret == PICOQUIC_ERROR_FRAME_BUFFER_TOO_SMALL) {
-                ret = 0;
-                *next_wake_time = current_time;
+            cnx->initial_repeat_needed = 0;
+            if (cnx->initial_validated) {
```

Comment:
Actually it boils down to this question: What exactly should happen once `initial_repeat_needed` was set to true?

My understanding from the spec would be a retry packet, but that doesn't seems to be the case.

Reply:
This is kind of explained in issue #344. The spec asks implementations to mitigate a security threat, at attack in which the Quic server is used to amplify a DDOS attack. The attack works as follow:

* attacker composes an initial packet (client hello), and sends it from the spoofed address of a target.
* server sees the initial packet, and reply with a first flight of typically 3 packets. The 3 packets are sent to the target.
* after a timer, the server notice that nothing was acked, so it repeats the packet.
* it might do that up to 4 times, sending 12 or 15 packets to the DDOS victim.

The mitigation is for the server to not acknowledge Initial packets before the address of the client is verified. That could be either because the server sent a Retry and the client responded with the proper token, or because the server sent an initial packet and the client responded with the server selected connection ID. This is tested in `packet.c:picoquic_incoming_client_initial()` and the result is stored in `cnx->initial_validated`.

That the server does not acknowledge the initial packets is not a big issue in most cases, because the client will receive the first flight, and the successful handshake is an implicit acknowledgement of the client initial. That's not an issue either if the client's initial packet is lost, because the client will automatically repeat it. If the server's packet is lost, the client will also repeat the client initial, and at that point the server must repeat it even if the connection is not yet verified. That's the condition noted in `cnx->initial_repeat_needed`.

Reply:
The reason for not sending ACK is the following failure: client sends initial, servers sends reply with ack, client sends ack, client's ack is lost. In that case, the client never repeats the initial because it was acked, and the server also never repeats its packets because of DDOS prevention. (Or we have to not implement DDOS prevention, which was open bug #344.)

Reply:
Thanks Christian for the nice explanation.

With that the code makes more sense to me now!

Reply:
Before checking in, I added a unit test to verify that the server does not send too much data in response to a lone client initial packet.